File: scripts/generate-xg-plus-data.mjs
import fs from 'node:fs';
import path from 'node:path';

const ROOT = process.cwd();
const SOURCE_FILE = path.resolve(ROOT, '../lab/projects/xG-plus/paul/tf/sample_1s_all.csv');
const OUT_DIR = path.resolve(ROOT, 'public/xg-plus');
const OUT_GIF_DIR = path.join(OUT_DIR, 'gifs');
const OUT_FILE = path.join(OUT_DIR, 'data.json');
const BBC_GIF_DIR = path.resolve(ROOT, '../lab/projects/xG-plus/bbc');

const MIN_TEAM_SHOTS = 150;
const MIN_ATTACKER_SHOTS = 20;

const GIFS = [
  { file: 'atk268.gif', caption: 'A cutback to Trent Alexander-Arnold creates the game-winning chance.' },
  { file: 'atk285.gif', caption: 'Harvey Elliott dribbles inside and curls the winner past three defenders.' },
  { file: 'atk340.gif', caption: 'Mohamed Salah dribbles inside and curls in a low-angle finish.' },
  { file: 'atk408.gif', caption: 'Jeremy Doku cuts inside and scores.' },
];

const TEAM_CANONICAL = {
  'AFC Bournemouth': 'Bournemouth',
  'West Ham': 'West Ham United',
};

function parseCSV(text) {
  const lines = text.trim().split(/\r?\n/);
  if (lines.length < 2) return [];
  const headers = parseLine(lines[0]);
  return lines.slice(1).map((line) => {
    const cols = parseLine(line);
    const row = {};
    headers.forEach((h, i) => {
      row[h] = cols[i] ?? '';
    });
    return row;
  });
}

function parseLine(line) {
  const out = [];
  let cur = '';
  let inQuotes = false;

  for (let i = 0; i < line.length; i += 1) {
    const ch = line[i];
    if (ch === '"') {
      if (inQuotes && line[i + 1] === '"') {
        cur += '"';
        i += 1;
      } else {
        inQuotes = !inQuotes;
      }
      continue;
    }
    if (ch === ',' && !inQuotes) {
      out.push(cur);
      cur = '';
      continue;
    }
    cur += ch;
  }

  out.push(cur);
  return out;
}

function toNum(v) {
  const n = Number(v);
  return Number.isFinite(n) ? n : null;
}

function toBool(v) {
  return String(v).toLowerCase() === 'true';
}

function canonicalTeam(name) {
  const trimmed = String(name || '').trim();
  return TEAM_CANONICAL[trimmed] ?? trimmed;
}

function seasonFromDate(dateStr) {
  const d = new Date(dateStr);
  if (!Number.isFinite(d.getTime())) return null;
  const year = d.getUTCFullYear();
  const month = d.getUTCMonth() + 1;
  if (month >= 7) return `${year}-${year + 1}`;
  return `${year - 1}-${year}`;
}

function startYearFromSeason(season) {
  const head = Number(String(season).slice(0, 4));
  return Number.isFinite(head) ? head : null;
}

function ensureModeBucket(map, season, mode) {
  if (!map.has(season)) map.set(season, { teams: new Map(), attackers: new Map() });
  return map.get(season)[mode];
}

function upsertTeam(bucket, name) {
  if (!bucket.has(name)) {
    bucket.set(name, {
      key: name,
      name,
      shotsFor: 0,
      xgPlusFor: 0,
      goalsFor: 0,
      shotsAgainst: 0,
      xgPlusAgainst: 0,
      goalsAgainst: 0,
    });
  }
  return bucket.get(name);
}

function upsertAttacker(bucket, id, name, team) {
  const cleanName = String(name || '').trim();
  const key = cleanName.toLowerCase();
  if (!bucket.has(key)) {
    bucket.set(key, {
      key,
      name: cleanName,
      playerId: id || null,
      shots: 0,
      xgPlusFor: 0,
      goals: 0,
      teams: new Set(),
    });
  }
  const row = bucket.get(key);
  if (team) row.teams.add(team);
  return row;
}

function percentileForRank(index, n) {
  if (n <= 1) return 50;
  return (index / (n - 1)) * 100;
}

function finalizeMode({ seasonBuckets, seasons, minShots, rawScore, extraFields = () => ({}) }) {
  const yearlyValues = new Map();

  for (const season of seasons) {
    const entries = seasonBuckets.get(season) || new Map();
    const accepted = [];

    for (const entity of entries.values()) {
      const shots = entity.shots ?? entity.shotsFor ?? 0;
      if (shots < minShots) continue;
      const raw = rawScore(entity);
      if (!Number.isFinite(raw)) continue;
      accepted.push({
        key: entity.key,
        name: entity.name,
        shots,
        raw,
        goals: entity.goals ?? entity.goalsFor ?? 0,
        teams: entity.teams ? [...entity.teams].sort((a, b) => a.localeCompare(b)) : undefined,
        ...extraFields(entity),
      });
    }

    const mean = accepted.length
      ? accepted.reduce((sum, x) => sum + x.raw, 0) / accepted.length
      : 0;
    const sorted = [...accepted].sort((a, b) => a.raw - b.raw);
    const rankMap = new Map();
    sorted.forEach((row, i) => rankMap.set(row.key, i));

    yearlyValues.set(
      season,
      accepted.map((row) => ({
        ...row,
        centered: row.raw - mean,
        percentile: percentileForRank(rankMap.get(row.key), sorted.length),
      })),
    );
  }

  const byEntity = new Map();

  for (const season of seasons) {
    const year = startYearFromSeason(season);
    const points = yearlyValues.get(season) || [];
    for (const p of points) {
      if (!byEntity.has(p.key)) {
        byEntity.set(p.key, {
          key: p.key,
          name: p.name,
          totalShots: 0,
          centeredSum: 0,
          seasons: new Set(),
          teams: new Set(),
          series: [],
        });
      }
      const entity = byEntity.get(p.key);
      entity.totalShots += p.shots;
      entity.centeredSum += p.centered;
      entity.seasons.add(season);
      (p.teams || []).forEach((team) => entity.teams.add(team));
      entity.series.push({
        season,
        year,
        raw: p.raw,
        centered: p.centered,
        percentile: p.percentile,
        shots: p.shots,
        goals: p.goals,
        forRate: p.forRate,
        againstRate: p.againstRate,
        diffRate: p.diffRate,
      });
    }
  }

  const players = [...byEntity.values()].map((entity) => {
    entity.series.sort((a, b) => a.year - b.year);
    const seasonsPresent = entity.seasons.size;
    const meanCentered = seasonsPresent ? entity.centeredSum / seasonsPresent : 0;
    return {
      name: entity.name,
      seasonsPresent,
      totalShots: entity.totalShots,
      meanCentered,
      teams: entity.teams.size ? [...entity.teams].sort((a, b) => a.localeCompare(b)) : undefined,
      series: entity.series,
    };
  });

  players.sort((a, b) => {
    if (b.meanCentered !== a.meanCentered) return b.meanCentered - a.meanCentered;
    if (b.seasonsPresent !== a.seasonsPresent) return b.seasonsPresent - a.seasonsPresent;
    if (b.totalShots !== a.totalShots) return b.totalShots - a.totalShots;
    return a.name.localeCompare(b.name);
  });

  return players;
}

function buildPayload() {
  const raw = fs.readFileSync(SOURCE_FILE, 'utf8');
  const rows = parseCSV(raw);
  const seasonData = new Map();
  const attackMap = new Map();

  for (const row of rows) {
    const season = seasonFromDate(row.date);
    const homeTeam = canonicalTeam(row.home_name);
    const awayTeam = canonicalTeam(row.away_name);
    const homeId = String(row.home_id || '').trim();
    const awayId = String(row.away_id || '').trim();
    const attackId = String(row.attack_team_id || '').trim();
    const attackMerged = String(row.attack_merged || '').trim();
    const gameId = String(row.game || '').trim();
    const playerName = String(row.player_name || '').trim();
    const playerId = String(row.player_id || '').trim();
    const goalProba = toNum(row.goal_proba) ?? 0;
    const isGoal = toBool(row.is_goal);
    const isShot = toBool(row.is_shot);

    if (!season || !homeTeam || !awayTeam || !homeId || !awayId || !attackId) continue;

    let attackTeam = null;
    let defenseTeam = null;
    if (attackId === homeId) {
      attackTeam = homeTeam;
      defenseTeam = awayTeam;
    } else if (attackId === awayId) {
      attackTeam = awayTeam;
      defenseTeam = homeTeam;
    } else {
      continue;
    }

    // Team-level xG+ should be one value per attack chain, not per 1-second sample.
    const chainId = attackMerged || `shot:${gameId}:${row.period}:${row.periodGameClockTime}:${playerId || playerName}`;
    if (chainId && gameId) {
      const attackKey = `${season}|${gameId}|${chainId}|${attackTeam}|${defenseTeam}`;
      const p = Math.max(0, Math.min(1, goalProba));
      if (!attackMap.has(attackKey)) {
        attackMap.set(attackKey, {
          season,
          attackTeam,
          defenseTeam,
          complementProd: 1 - p,
          isGoal,
        });
      } else {
        const bucket = attackMap.get(attackKey);
        bucket.complementProd *= 1 - p;
        if (isGoal) bucket.isGoal = true;
      }
    }

    // Keep attacker aggregation at shot level using shooter attribution.
    if (isShot && playerName) {
      const attackerBucket = ensureModeBucket(seasonData, season, 'attackers');
      const attacker = upsertAttacker(attackerBucket, playerId, playerName, attackTeam);
      attacker.shots += 1;
      attacker.xgPlusFor += goalProba;
      if (isGoal) attacker.goals += 1;
    }
  }

  for (const attack of attackMap.values()) {
    const teamBucket = ensureModeBucket(seasonData, attack.season, 'teams');
    const attackTeamRow = upsertTeam(teamBucket, attack.attackTeam);
    const defenseTeamRow = upsertTeam(teamBucket, attack.defenseTeam);
    const chainProb = Math.max(0, Math.min(1, 1 - attack.complementProd));

    attackTeamRow.shotsFor += 1;
    attackTeamRow.xgPlusFor += chainProb;
    if (attack.isGoal) attackTeamRow.goalsFor += 1;

    defenseTeamRow.shotsAgainst += 1;
    defenseTeamRow.xgPlusAgainst += chainProb;
    if (attack.isGoal) defenseTeamRow.goalsAgainst += 1;
  }

  const seasons = [...seasonData.keys()].sort((a, b) => a.localeCompare(b));
  const teamPlayers = finalizeMode({
    seasonBuckets: new Map(seasons.map((s) => [s, seasonData.get(s).teams])),
    seasons,
    minShots: MIN_TEAM_SHOTS,
    rawScore: (t) => t.xgPlusFor - t.xgPlusAgainst,
    extraFields: (t) => {
      const forRate = t.xgPlusFor;
      const againstRate = t.xgPlusAgainst;
      return {
        forRate,
        againstRate,
        diffRate: forRate - againstRate,
      };
    },
  });

  const attackerPlayers = finalizeMode({
    seasonBuckets: new Map(seasons.map((s) => [s, seasonData.get(s).attackers])),
    seasons,
    minShots: MIN_ATTACKER_SHOTS,
    rawScore: (a) => a.xgPlusFor,
    extraFields: (a) => {
      const forRate = a.xgPlusFor;
      return {
        forRate,
        againstRate: null,
        diffRate: null,
      };
    },
  });

  return {
    generatedAt: new Date().toISOString(),
    dataSource: {
      file: 'lab/projects/xG-plus/paul/tf/sample_1s_all.csv',
      competition: 'EPL',
      measure: 'teams: attack-level cumulative 1 - product(1 - goal_proba) from 1-second samples; attackers: shot-level goal_proba',
      seasonRule: 'july_boundary',
    },
    seasons,
    metricLabels: {
      centered: 'Centered xG+',
      percentile: 'xG+ Percentile',
    },
    gifs: GIFS.map((g) => ({
      file: g.file,
      src: `xg-plus/gifs/${g.file}`,
      caption: g.caption,
    })),
    modes: {
      teams: {
        label: 'Teams',
        minShotsPerSeason: MIN_TEAM_SHOTS,
        defaultPlayers: teamPlayers.slice(0, 5).map((p) => p.name),
        players: teamPlayers,
      },
      attackers: {
        label: 'Attackers',
        minShotsPerSeason: MIN_ATTACKER_SHOTS,
        defaultPlayers: attackerPlayers.slice(0, 5).map((p) => p.name),
        players: attackerPlayers,
      },
    },
  };
}

function copyGifs() {
  fs.mkdirSync(OUT_GIF_DIR, { recursive: true });
  for (const gif of GIFS) {
    const src = path.join(BBC_GIF_DIR, gif.file);
    const dest = path.join(OUT_GIF_DIR, gif.file);
    if (!fs.existsSync(src)) {
      throw new Error(`Missing GIF source: ${src}`);
    }
    fs.copyFileSync(src, dest);
  }
}

function main() {
  if (!fs.existsSync(SOURCE_FILE)) {
    throw new Error(`Missing source file: ${SOURCE_FILE}`);
  }
  fs.mkdirSync(OUT_DIR, { recursive: true });
  copyGifs();
  const payload = buildPayload();
  fs.writeFileSync(OUT_FILE, `${JSON.stringify(payload)}\n`, 'utf8');
  console.log(`[xg-plus] Wrote ${OUT_FILE}`);
}

main();

File: scripts/generate-xg-plus-data.py
from __future__ import annotations

import json
import math
import shutil
from pathlib import Path

import numpy as np
import pandas as pd


ROOT = Path(__file__).resolve().parents[1]
SOURCE_FILE = ROOT.parent / "lab/projects/xG-plus/paul/tf/pl_2022-2025.parquet"
OUT_DIR = ROOT / "public/xg-plus"
OUT_GIF_DIR = OUT_DIR / "gifs"
OUT_FILE = OUT_DIR / "data.json"
BBC_GIF_DIR = ROOT.parent / "lab/projects/xG-plus/bbc"

MIN_TEAM_CHAINS = 150
MIN_ATTACKER_CHANCES = 20

GIFS = [
    {"file": "atk268.gif", "caption": "A cutback to Trent Alexander-Arnold creates the game-winning chance."},
    {"file": "atk285.gif", "caption": "Harvey Elliott dribbles inside and curls the winner past three defenders."},
    {"file": "atk340.gif", "caption": "Mohamed Salah dribbles inside and curls in a low-angle finish."},
    {"file": "atk408.gif", "caption": "Jeremy Doku cuts inside and scores."},
]

TEAM_CANONICAL = {
    "AFC Bournemouth": "Bournemouth",
    "West Ham": "West Ham United",
}


def canonical_team(name: str) -> str:
    n = (name or "").strip()
    return TEAM_CANONICAL.get(n, n)


def season_from_date(date_series: pd.Series) -> pd.Series:
    dt = pd.to_datetime(date_series, errors="coerce")
    year = dt.dt.year
    month = dt.dt.month
    start_year = np.where(month >= 7, year, year - 1)
    return pd.Series(start_year.astype(str) + "-" + (start_year + 1).astype(str), index=date_series.index)


def percentile_from_rank(rank: int, n: int) -> float:
    if n <= 1:
        return 50.0
    return (rank / (n - 1)) * 100.0


def finalize_entities(rows_by_season: dict[str, list[dict]], min_count: int) -> list[dict]:
    by_entity: dict[str, dict] = {}

    for season in sorted(rows_by_season.keys()):
        rows = [r for r in rows_by_season[season] if r["shots"] >= min_count and math.isfinite(r["raw"])]
        if not rows:
            continue

        mean_raw = sum(r["raw"] for r in rows) / len(rows)
        sorted_rows = sorted(rows, key=lambda x: x["raw"])
        rank_map = {r["key"]: i for i, r in enumerate(sorted_rows)}

        for r in rows:
            key = r["key"]
            centered = r["raw"] - mean_raw
            percentile = percentile_from_rank(rank_map[key], len(sorted_rows))
            season_year = int(season[:4])

            if key not in by_entity:
                by_entity[key] = {
                    "name": r["name"],
                    "totalShots": 0,
                    "centeredSum": 0.0,
                    "seasons": set(),
                    "teams": set(),
                    "series": [],
                }

            e = by_entity[key]
            e["totalShots"] += int(r["shots"])
            e["centeredSum"] += centered
            e["seasons"].add(season)
            for t in r.get("teams", []):
                e["teams"].add(t)
            e["series"].append(
                {
                    "season": season,
                    "year": season_year,
                    "raw": r["raw"],
                    "centered": centered,
                    "percentile": percentile,
                    "shots": int(r["shots"]),
                    "goals": int(r["goals"]),
                    "forRate": r.get("forRate"),
                    "againstRate": r.get("againstRate"),
                    "diffRate": r.get("diffRate"),
                }
            )

    out = []
    for e in by_entity.values():
        e["series"].sort(key=lambda x: x["year"])
        seasons_present = len(e["seasons"])
        mean_centered = e["centeredSum"] / seasons_present if seasons_present else 0.0
        out.append(
            {
                "name": e["name"],
                "seasonsPresent": seasons_present,
                "totalShots": e["totalShots"],
                "meanCentered": mean_centered,
                "teams": sorted(e["teams"]) if e["teams"] else None,
                "series": e["series"],
            }
        )

    out.sort(key=lambda x: (-x["meanCentered"], -x["seasonsPresent"], -x["totalShots"], x["name"]))
    return out


def build_payload() -> dict:
    cols = [
        "game",
        "date",
        "home_id",
        "home_name",
        "away_id",
        "away_name",
        "attack_team_id",
        "attack_merged",
        "period",
        "periodGameClockTime",
        "player_id",
        "player_name",
        "is_shot",
        "goal_proba",
        "is_goal",
    ]
    df = pd.read_parquet(SOURCE_FILE, columns=cols)

    df["home_name"] = df["home_name"].astype(str).map(canonical_team)
    df["away_name"] = df["away_name"].astype(str).map(canonical_team)
    df["season"] = season_from_date(df["date"])

    df["home_id"] = df["home_id"].astype(str)
    df["away_id"] = df["away_id"].astype(str)
    df["attack_team_id"] = df["attack_team_id"].astype(str)
    home_mask = df["attack_team_id"] == df["home_id"]
    away_mask = df["attack_team_id"] == df["away_id"]
    df = df[home_mask | away_mask].copy()

    df["attack_team"] = np.where(home_mask.loc[df.index], df["home_name"], df["away_name"])
    df["defense_team"] = np.where(home_mask.loc[df.index], df["away_name"], df["home_name"])

    df["goal_proba"] = pd.to_numeric(df["goal_proba"], errors="coerce").fillna(0.0).clip(0, 1)
    df["is_goal"] = df["is_goal"].astype(bool)
    df["is_shot"] = df["is_shot"].astype(bool)

    # second bucket consolidation (to avoid next-second double counting artifacts)
    df["int_sec"] = pd.to_numeric(df["periodGameClockTime"], errors="coerce").fillna(0).astype(int)
    df["attack_merged"] = df["attack_merged"].fillna(-1).astype(int).astype(str)
    df["chain_key"] = df["game"].astype(str) + "|" + df["attack_merged"]

    sec = (
        df.groupby(
            ["season", "game", "attack_team", "defense_team", "chain_key", "int_sec"], as_index=False
        )
        .agg(sec_p=("goal_proba", "max"), sec_goal=("is_goal", "max"))
    )
    sec["log_comp"] = np.log1p(-sec["sec_p"].clip(0, 1))

    chain = (
        sec.groupby(["season", "game", "attack_team", "defense_team", "chain_key"], as_index=False)
        .agg(sum_log_comp=("log_comp", "sum"), chain_goal=("sec_goal", "max"), chain_secs=("int_sec", "nunique"))
    )
    chain["chain_p"] = 1 - np.exp(chain["sum_log_comp"])
    total_pred = float(chain["chain_p"].sum())
    total_goals = float(chain["chain_goal"].sum())
    correction_factor = (total_goals / total_pred) if total_pred > 0 else 1.0

    for_team = (
        chain.groupby(["season", "attack_team"], as_index=False)
        .agg(shotsFor=("chain_key", "nunique"), xgPlusFor=("chain_p", "sum"), goalsFor=("chain_goal", "sum"))
        .rename(columns={"attack_team": "team"})
    )
    against_team = (
        chain.groupby(["season", "defense_team"], as_index=False)
        .agg(
            shotsAgainst=("chain_key", "nunique"),
            xgPlusAgainst=("chain_p", "sum"),
            goalsAgainst=("chain_goal", "sum"),
        )
        .rename(columns={"defense_team": "team"})
    )
    team = for_team.merge(against_team, how="outer", on=["season", "team"]).fillna(0)
    team["xgPlusFor"] = team["xgPlusFor"] * correction_factor
    team["xgPlusAgainst"] = team["xgPlusAgainst"] * correction_factor
    team["diff"] = team["xgPlusFor"] - team["xgPlusAgainst"]

    # Attacker view (chance-level cumulative xG+ by player)
    players = df.copy()
    players["player_name"] = players["player_name"].astype(str).str.strip()
    players = players[players["player_name"] != ""]

    sec_player = (
        players.groupby(
            ["season", "player_name", "game", "chain_key", "int_sec"], as_index=False
        )
        .agg(sec_p=("goal_proba", "max"), sec_goal=("is_goal", "max"))
    )
    sec_player["log_comp"] = np.log1p(-sec_player["sec_p"].clip(0, 1))

    chain_player = (
        sec_player.groupby(["season", "player_name", "game", "chain_key"], as_index=False)
        .agg(sum_log_comp=("log_comp", "sum"), chance_secs=("int_sec", "nunique"))
    )
    chain_player["chain_p"] = 1 - np.exp(chain_player["sum_log_comp"])
    chain_player["chain_p"] = chain_player["chain_p"] * correction_factor

    att = (
        chain_player.groupby(["season", "player_name"], as_index=False)
        .agg(chances=("chain_key", "nunique"), xgPlusFor=("chain_p", "sum"))
        .rename(columns={"player_name": "name"})
    )
    goals_by_player = (
        players[players["is_goal"]]
        .groupby(["season", "player_name"], as_index=False)
        .size()
        .rename(columns={"player_name": "name", "size": "goals"})
    )
    att = att.merge(goals_by_player, on=["season", "name"], how="left")
    att["goals"] = att["goals"].fillna(0).astype(int)

    att_teams = (
        players.groupby(["season", "player_name"])["attack_team"]
        .agg(lambda x: sorted(set(x)))
        .reset_index()
        .rename(columns={"player_name": "name", "attack_team": "teams"})
    )
    att = att.merge(att_teams, on=["season", "name"], how="left")

    seasons = sorted(df["season"].dropna().unique().tolist())

    team_rows_by_season: dict[str, list[dict]] = {s: [] for s in seasons}
    for _, r in team.iterrows():
        season = r["season"]
        team_rows_by_season[season].append(
            {
                "key": r["team"],
                "name": r["team"],
                "shots": int(r["shotsFor"]),
                "raw": float(r["diff"]),
                "goals": int(r["goalsFor"]),
                "forRate": float(r["xgPlusFor"]),
                "againstRate": float(r["xgPlusAgainst"]),
                "diffRate": float(r["diff"]),
            }
        )

    att_rows_by_season: dict[str, list[dict]] = {s: [] for s in seasons}
    for _, r in att.iterrows():
        season = r["season"]
        att_rows_by_season[season].append(
            {
                "key": str(r["name"]).strip().lower(),
                "name": r["name"],
                "shots": int(r["chances"]),
                "raw": float(r["xgPlusFor"]),
                "goals": int(r["goals"]),
                "forRate": float(r["xgPlusFor"]),
                "againstRate": None,
                "diffRate": None,
                "teams": r["teams"] if isinstance(r["teams"], list) else [],
            }
        )

    team_players = finalize_entities(team_rows_by_season, MIN_TEAM_CHAINS)
    attacker_players = finalize_entities(att_rows_by_season, MIN_ATTACKER_CHANCES)

    return {
        "generatedAt": pd.Timestamp.utcnow().isoformat(),
        "dataSource": {
            "file": "lab/projects/xG-plus/paul/tf/pl_2022-2025.parquet",
            "competition": "EPL",
            "measure": "teams: second-bucketed chain cumulative 1 - product(1 - p); attackers: player-chain cumulative 1 - product(1 - p)",
            "seasonRule": "july_boundary",
            "visualScaleCorrectionFactor": correction_factor,
            "visualScaleCorrectionBasis": "global_goals_over_global_predicted_chain_sum",
        },
        "seasons": seasons,
        "metricLabels": {
            "centered": "Centered xG+",
            "percentile": "xG+ Percentile",
        },
        "gifs": [{"file": g["file"], "src": f"xg-plus/gifs/{g['file']}", "caption": g["caption"]} for g in GIFS],
        "modes": {
            "teams": {
                "label": "Teams",
                "minShotsPerSeason": MIN_TEAM_CHAINS,
                "defaultPlayers": [p["name"] for p in team_players[:5]],
                "players": team_players,
            },
            "attackers": {
                "label": "Attackers",
                "minShotsPerSeason": MIN_ATTACKER_CHANCES,
                "defaultPlayers": [p["name"] for p in attacker_players[:5]],
                "players": attacker_players,
            },
        },
    }


def copy_gifs() -> None:
    OUT_GIF_DIR.mkdir(parents=True, exist_ok=True)
    for gif in GIFS:
        src = BBC_GIF_DIR / gif["file"]
        if not src.exists():
            raise FileNotFoundError(f"Missing GIF source: {src}")
        shutil.copy2(src, OUT_GIF_DIR / gif["file"])


def main() -> None:
    if not SOURCE_FILE.exists():
        raise FileNotFoundError(f"Missing source file: {SOURCE_FILE}")
    OUT_DIR.mkdir(parents=True, exist_ok=True)
    copy_gifs()
    payload = build_payload()
    OUT_FILE.write_text(json.dumps(payload) + "\n", encoding="utf-8")
    print(f"[xg-plus] Wrote {OUT_FILE}")


if __name__ == "__main__":
    main()
